File: YunChat-Client/src/service/impl/EncryptServiceImpl.ts
import CryptoJS from 'crypto-js';
import config from '@/socket/config';
import { EncryptService } from './../EncryptService';
import { deviceServiceInstance } from '../ServiceInstance';

/**
 * @name 加密服务实现类
 * @author yichen9247
 */
export class EncryptServiceImpl implements EncryptService {
    private readonly SECRET_KEY: string = config.server.config.secretKey;

    /**
     * @name 加密文本
     * @param plainText 待加密的文本
     * @returns Base64格式的加密字符串
     */
    private encryptText(plainText: string): string {
        return CryptoJS.AES.encrypt(plainText, CryptoJS.enc.Utf8.parse(this.SECRET_KEY), {
            mode: CryptoJS.mode.CBC,
            padding: CryptoJS.pad.Pkcs7,
            iv: CryptoJS.enc.Utf8.parse(this.SECRET_KEY)
        }).toString();
    }

    /**
     * @name 加密授权信息
     * @returns Promise<string> Base64格式的加密字符串
     */
    async encryptAuthorization(): Promise<string> {
        const timestamp = Date.now();
        const uid = localStorage.getItem("yun_uid") || "";
        const token = localStorage.getItem("yun_token") || "";
        const isDesktop = deviceServiceInstance.getDeviceIsDesktop();
        const plainText = `${uid}||${token}||${isDesktop ? "Desktop" : "Web"}||${timestamp}`;
        return this.encryptText(plainText);
    }

    /**
     * @name 加密登录信息
     * @returns Promise<string> Base64格式的加密字符串
     */
    async encryptLogin(username: string, password: string): Promise<string> {
        const timestamp = Date.now();
        const isDesktop = deviceServiceInstance.getDeviceIsDesktop();
        const plainText = `${username}||${password}||${isDesktop ? "Desktop" : "Web"}||${timestamp}`;
        return this.encryptText(plainText);
    }
}

File: YunChat-Android/app/src/main/kotlin/com/android/yunchat/service/impl/EncryptServiceImpl.kt
package com.android.yunchat.service.impl

import android.util.Base64
import com.android.yunchat.config.AppConfig
import com.android.yunchat.service.EncryptService
import com.android.yunchat.service.instance.userServiceInstance
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * @name 加密服务实现类
 * @author yichen9247
 */
class EncryptServiceImpl: EncryptService {
    /**
     * @name 加密文本
     * @param plainText 待加密的文本
     * @returns Promise<string> Base64格式的加密字符串
     */
    private fun encryptText(plainText: String): String {
        val keySpec = SecretKeySpec(SECRET_KEY.toByteArray(Charsets.UTF_8), ALGORITHM)
        val ivSpec = IvParameterSpec(SECRET_KEY.substring(0, 16).toByteArray(Charsets.UTF_8))
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec)
        val encryptedBytes = cipher.doFinal(plainText.toByteArray(Charsets.UTF_8))
        return Base64.encodeToString(encryptedBytes, Base64.NO_WRAP)
    }

    /**
     * @name 加密授权信息
     * @returns Promise<string> Base64格式的加密字符串
     */
    override fun encryptAuthorization(): String {
        val uid = userServiceInstance.getUserUid()
        val token = userServiceInstance.getUserToken()
        val timestamp = System.currentTimeMillis()
        val plainText = "$uid||$token||App||$timestamp"
        return encryptText(plainText)
    }

    /**
     * @name 加密登录信息
     * @param username 账号
     * @param password 密码
     * @returns Promise<string> Base64格式的加密字符串
     */
    override fun encryptLogin(username: String, password: String): String {
        val timestamp = System.currentTimeMillis()
        val plainText = "$username||$password||App||$timestamp"
        return encryptText(plainText)
    }

    companion object {
        private const val ALGORITHM = "AES"
        private const val SECRET_KEY = AppConfig.secretKey
        private const val TRANSFORMATION = "AES/CBC/PKCS7Padding"
    }
}